Bump supported php versions

gennadigennadigennadi opened the pull request:
Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| QA            | yes

### Description
Update minimum PHP Version to 7.4. I also updated to PHPunit 9 and fixed the deprecations. All these update are relevant for making laminas-di compatible with PHP 8, since only PHPunit 9 is going to support it.

I have also notices that lamians-servicemanager 3.5 will move in AbstractFactoryInterface to Psr\Container\ContainerInterface, that's why laminas-di will break for now.

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

gennadigennadigennadi committed:
bump php version

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

boesing commented:
@gennadigennadigennadi First of all, thanks for your contribution.
Would be great if you include PHP 7.3 as its still supported until end of december 2020.
Please also cleanup `.travis.yml` so it does not test against PHP 7.1 + PHP 7.2.

If you would like to help us by adding PHP 8.0 aswell, please have a look at this [gist](https://gist.github.com/boesing/11ce14a36cd42dda547b79877fc074f2). 👍

gennadigennadigennadi committed:
update minimum php version to 7.3 and update .travis.yml

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

gennadigennadigennadi committed:
reset phpstan/phpstan-shim
because phpstan 0.12 is more strict and fixing can be done later

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

gennadigennadigennadi committed:
updated phpunit.xml.dist

--migrate-configuration

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

gennadigennadigennadi committed:
add php 8

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

gennadigennadigennadi committed:
replace phpstan-shim with phpstan

- phpstan-shim replaces nikic/php-parser and therefore test-coverage fails
- phpstan 0.12 changes the levels so i had to lower the level

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

gennadigennadigennadi committed:
replace phpstan-shim with phpstan

- phpstan-shim replaces nikic/php-parser and therefore test-coverage fails
- phpstan 0.12 changes the levels so i had to lower the level
- added phpstan-baseline

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

boesing commented on .travis.yml:
```diff
@@ -12,32 +12,28 @@ env:
+    - php: nightly
       env:
         - DEPS=latest
+  allow_failures:
```
Please do not allow failures for PHP 8.0 as it is in RC now

boesing commented on .travis.yml:
```diff
@@ -12,32 +12,28 @@ env:
-    - php: 7.4
+    - php: nightly
       env:
         - DEPS=latest
```
for nightly, you can add `COMPOSER_ARGUMENTS="$COMPOSER_ARGUMENTS --ignore-platform-reqs` so it can install all dependencies even if these are not yet supporting PHP 8.0.

gennadigennadigennadi replied:
it's already set for all php version: https://github.com/laminas/laminas-di/blob/3.2.x/.travis.yml#L46

boesing replied:
As you can see, build on PHP 8 is still failing. This is due to the fact that the `COMPOSER_ARGUMENTS` do not gobally ignore platform reqs. To fix this, you can drop `--ignore-platform-reqs` from line 46 and instead, add it to the nightly envs. After this, travis should pass.

boesing commented on phpstan.neon:
```diff
@@ -1,5 +1,5 @@
 parameters:
-    level: 7
+    level: 4
```
better create a baseline or drop phpstan at all as we are heading to psalm anyways 👍

gennadigennadigennadi committed:
remove allow_failures for nightly

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

gennadigennadigennadi committed:
add ignore-platform-reqs for nightlies

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

boesing commented:
Oh, I was a bit too quick. Could you fix the unit test where the `array` error is popping up? I guess a simple `is_object || is_string` check would solve the problem with `method_exists` tho. 

@weierophinney @tux-rampage do you guys know why `__set_state` is static & private in that [UnexporatableValue2](https://github.com/laminas/laminas-di/blob/3.2.x/test/TestAsset/Resolver/UnexportableValue2.php) class?

```
class UnexportableValue2
{
    private static function __set_state($data)
    {
    }
}
```

tux-rampage commented:
> @weierophinney @tux-rampage do you guys know why `__set_state` is static & private in that [UnexporatableValue2](https://github.com/laminas/laminas-di/blob/3.2.x/test/TestAsset/Resolver/UnexportableValue2.php) class?
> 
> ```
> class UnexportableValue2
> {
>     private static function __set_state($data)
>     {
>     }
> }
> ```

Because it's representing a case where an instance of this class cannot be safely exported with `var_export()`. In this case the code generator has to detect this and avoid creating code that breaks with an error when executed.

I'll review the changes within the next days.

gennadigennadigennadi committed:
bump laminas/laminas/stdlib to ^3.3

we can safely drop support for php5.6

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

boesing commented:
> Because it's representing a case where an instance of this class cannot be safely exported with var_export()

Yes, because its even invalid PHP code. Sadly, PHP 7 never had a problem with marking that method as private. Not even sure if there is any magic method which can be private. 🤔

gennadigennadigennadi commented:
> Oh, I was a bit too quick. Could you fix the unit test where the `array` error is popping up? I guess a simple `is_object || is_string` check would solve the problem with `method_exists` tho.
> 

The Problem was already fixed in laminas-stdlib 3.3, therefore I just bump the minimal version.

tux-rampage commented:
> > Because it's representing a case where an instance of this class cannot be safely exported with var_export()
> 
> Yes, because its even invalid PHP code. Sadly, PHP 7 never had a problem with marking that method as private. Not even sure if there is any magic method which can be private. thinking

That's why it's there. If people *can* do it some *will* do it. I don't mind dropping this test case and stick with `resource` as non-exportable. For everything else we could refer to the PHP docs and suggest tu use static analysis.

As for static analysis: Psalm with full Template support and strictest level would be my next goal for this component - before improving the AoT generator and getting rid of some code that is no longer necessary.

tux-rampage reviewed (changes requested):
@gennadigennadigennadi Thank you for your contribution. :+1:

Would you mind taking care of removing the [`UnexeportableValue2`](/laminas/laminas-di/blob/3.2.x/test/TestAsset/Resolver/UnexportableValue2.php) and it's references in the test cases (Should only be in the provider of the [`ValueInjectionTest`](https://github.com/laminas/laminas-di/blob/d909fbc5a48ac5dbc93a01e0e03378d65c0dd2df/test/Resolver/ValueInjectionTest.php#L124) so we get rid of the build failure?

Otherwise it looks good so far. Thank you very much.

gennadigennadigennadi committed:
remove UnexportableValue2.php

Signed-off-by: gennadigennadigennadi <gennadijanzen@icloud.com>

tux-rampage commented on test/Resolver/ValueInjectionTest.php:
```diff
@@ -27,13 +28,15 @@
         parent::setUp();
 
-        if (! $this->streamFixture) {
+        if (!$this->streamFixture) {
```
There should be a space between `!` and the condition.

tux-rampage commented on test/Resolver/ValueInjectionTest.php:
```diff
@@ -123,16 +126,14 @@ public function provideExportableValues()
     public function provideUnexportableItems()
     {
-        if (! $this->streamFixture) {
+        if (!$this->streamFixture) {
```
Same as above: There should be a space between `!` and the condition.

tux-rampage reviewed (changes requested):
Thank you very much. The last commit seemed to introduce some style violations by accident.

Commit:
add space after !

Signed-off-by: Gennadi Janzen <gennadi.janzen@check24.de>

tux-rampage reviewed (approved):
Well done. LGTM
